### PR TITLE
EOS-16372: mini provisioner changes for 1,3 node deployment

### PR DESCRIPTION
**Pull request by mariyappanp**

### CODE REVIEW

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        # Validate BMC & Storage controller accessibility
+        if node_type.lower() != "vm":
+            BmcV().validate('accessible', [socket.getfqdn(), bmc_ip, bmc_user, bmc_passwd])
+            c_validator = ControllerV()
+            c_validator.validate("accessible", [primary_ip, cntrlr_user, cntrlr_passwd])
+            c_validator.validate("accessible", [secondary_ip, cntrlr_user, cntrlr_passwd])
+
+            # Validate network fqdn reachability
+            # mgmt public fqdn should also be for VM?
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:101]: Need to check with CSM

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    def validate_nw_interface(self, interfaces):
+        """Check network interfaces are up.
+        carrier:0 - link down
+        carrier:1 - link up
+        """
+        if not isinstance(interfaces, list):
+            raise SetupError(errno.EINVAL, "%s - validation failure. %s",
+                self.name, "Expected list of interfaces. Received, %s." % interfaces)
+        for interface in interfaces:
+            cmd = "cat /sys/class/net/%s/carrier" % interface
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:122]: I think 'carrier' was for cable. Additionally can use 'ip' command & check interface reachability

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py [hunk 3]
```diff
@@ -89,46 +91,67 @@ def add_args(parser: str, cls: str, name: str):
         parsers.add_argument('args', nargs='*', default=[], help='args')
         parsers.set_defaults(command=cls)
 
+    def copy_provisioner_config(self, stage=None):
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:94]: rename to copy_input_config()

File: low-level/framework/base/sspl_constants.py [hunk 4]
```diff
@@ -33,13 +32,13 @@
 cs_products = ["CS-A"]
 cs_legacy_products = ["CS-L", "CS-G"]
 setups = ["vm", "cortx", "ssu", "gw", "cmu"]
-RESOURCE_PATH = "/opt/seagate/%s/sspl/low-level/json_msgs/schemas/" % (PRODUCT_FAMILY)
-CLI_RESOURCE_PATH = "/opt/seagate/%s/sspl/low-level/tests/manual" % (PRODUCT_FAMILY)
-DATA_PATH = "/var/%s/sspl/data/" % (PRODUCT_FAMILY)
-SSPL_CONFIGURED_DIR = "/var/%s/sspl" % (PRODUCT_FAMILY)
+RESOURCE_PATH = "/opt/seagate/%s/sspl/low-level/json_msgs/schemas/" % PRODUCT_FAMILY
```
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:35]: Not sure why these changes even, existing ones are more readable anyways

File: low-level/sspl-ll.spec [hunk 5]
```diff
@@ -133,10 +110,6 @@ SSPL_DIR=/opt/seagate/%{product_family}/sspl
 # [ -f /etc/rsyslog.d/1-ssplfwd.conf ] ||
 #    cp /opt/seagate/%{product_family}/sspl/low-level/files/etc/rsyslog.d/1-ssplfwd.conf /etc/rsyslog.d/1-ssplfwd.conf
 
-# Copy init script
-[ -f $SSPL_DIR/sspl_init ] ||
```
Comment (cdeshmukh) [inline low-level/sspl-ll.spec:137]: Just make sure all these omitted steps are covered in relevant mini provisioner stage, including this one

**Review by gargpalak048 — COMMENTED** [review]
The default values for each key can be kept in sspl_constants.py file.
eg:- DEFAULT_SITE_ID = 'DC01'

**Review by gargpalak048 — COMMENTED** [review]
In each sensor, rather than making Conf apis calls for every key field,
Can't we have a function fetch_config() in conf_utils.. that function can accept the required key as a param.

**Review by gargpalak048 — COMMENTED** [review]
same for Conf.set api call, we can have a func set_config() in conf_utils?

**Review by gargpalak048 — COMMENTED** [review]
In file low-level/files/opt/seagate/sspl/setup/sspl_prepare.py
rather than passing the complete string to get_config_value(),
we can pass the required key and can map the key with the required str.
Also we can create a list of req_fields like ['cluster_id', 'site_id' ] and can pass to the get_config_value() func rather than calling the func again and again.

File: low-level/files/opt/seagate/sspl/conf/setup.yaml [hunk 6]
```diff
@@ -13,50 +13,54 @@
 # about this software or licensing, please email opensource@seagate.com or
 # cortx-questions@seagate.com.
 
-sspl: 
-  post_install: 
+sspl:
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/setup.yaml:16]: Earlier in main branch the sspl conf file names were like, sspl.conf.LR2.yaml and based on product value, we selected the conf. As per the the recommended naming convention on the conf/template files, now they don't have product name on it. Likely the sspl service file. With main branch, LR2 is default on the conf, service files. 

If still it is required to be, then conf files also should be named with LR2. Please confirm to proceed.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 1 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:101]: For VM, CSM adds mgmt public fqdn as localhost for Dev purpose.
Now this validation is kept for VM and HW. 
RE/CICD will use localhost or valid FQDN to avoid validation failure.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 2 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:122]: Okay

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py (hunk 3 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:94]: Okay

File: low-level/framework/base/sspl_constants.py (hunk 4 above)
Comment (mariyappanp) [reply low-level/framework/base/sspl_constants.py:35]: Just changed these lines to avoid parenthesis. Added back as you feel readability is better with parenthesis.

File: low-level/sspl-ll.spec (hunk 5 above)
Comment (mariyappanp) [reply low-level/sspl-ll.spec:137]: We don't have any use on sspl_init with main branch. 
keeping this as low priority at this time. I will add this once again get confirmed.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.yaml [hunk 7]
```diff
@@ -224,3 +224,5 @@ SYSTEM_INFORMATION:
    sspl_log_file_path: /var/log/cortx/sspl/sspl.log
    log_level: INFO
    environment: PROD
+   syslog_host: localhost
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.yaml:227]: No, provisioner conf store keys don't have it. 
This is under sspl.conf for managing syslog configuration.

File: low-level/files/opt/seagate/sspl/conf/setup.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/setup.yaml:16]: Reverted.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 2 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:122]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py (hunk 3 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py]: Done.

File: low-level/framework/base/sspl_constants.py (hunk 4 above)
Comment (mariyappanp) [reply low-level/framework/base/sspl_constants.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/generate_resource_health_view/resource_health_view [hunk 8]
```diff
@@ -38,9 +38,9 @@ from framework.utils.ipmi_client import IpmiFactory
 from framework.utils.sysfs_interface import SysFS
 from framework.utils.tool_factory import ToolFactory
 from framework.base.sspl_constants import PRODUCT_FAMILY , DATA_PATH
-from framework.utils.conf_utils import (CLUSTER, ENCLOSURE, GLOBAL_CONF, SRVNODE,
-    STORAGE, Conf, NODE_TYPE, TYPE, SITE_ID, RACK_ID, NODE_ID, CLUSTER_ID, RELEASE,
-    TARGET_BUILD)
+from framework.utils.conf_utils import (GLOBAL_CONF, SRVNODE,
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/generate_resource_health_view/resource_health_view:41]: TYPE not used.

File: low-level/files/opt/seagate/sspl/setup/sspl_config.py [hunk 9]
```diff
@@ -27,186 +27,96 @@
...
-                errno.EINVAL,
-                "Run this command with root privileges!!")
-
+        if os.geteuid() != 0:
+            raise SetupError(errno.EINVAL,
+                "Run this command with root privileges.")
+
+        # Validate input/provisioner configs
+        machine_id = Utility.get_machine_id()
+        cluster_id = Utility.get_config_value(consts.PRVSNR_CONFIG_INDEX,
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/sspl_config.py:66]: What is the use of this variable?

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py [hunk 10]
```diff
@@ -20,72 +20,92 @@
 ###################################################################
 
 import os
+import pwd
+import grp
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/sspl_post_install.py:24]: Not used in this module.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py [hunk 11]
```diff
@@ -20,72 +20,92 @@
...
+import errno
 import shutil
+import socket
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/sspl_post_install.py:27]: Not used in this module.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py [hunk 12]
```diff
@@ -20,72 +20,92 @@
...
 import distutils.dir_util
 from urllib.parse import urlparse
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/sspl_post_install.py:29]: not used in this module.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+# with this program. If not, see <https://www.gnu.org/licenses/>. For any questions
+# about this software or licensing, please email opensource@seagate.com or
+# cortx-questions@seagate.com.
+
+import os
+import errno
+import socket
+
+# using cortx package
+from cortx.utils.conf_store import Conf
```
Comment (thavanathan) [inline low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:23]: Not used.

**Review by thavanathan — COMMENTED** [review]
Some minor comments.

**Review by gargpalak048 — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml [hunk 14]
```diff
@@ -224,3 +224,5 @@ SYSTEM_INFORMATION:
    sspl_log_file_path: /var/log/cortx/sspl/sspl.log
    log_level: INFO
    environment: PROD
+   syslog_host: localhost
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml:227]: please remove syslog conf, no remote logging enabled in Cortxv2, so we this is not needed. Code can hardcode to this host & port

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py [hunk 15]
```diff
@@ -20,72 +20,92 @@
...
+        self.state_file = "%s/state.txt" % consts.DATA_PATH
+        self.dbus_service = DbusServiceHandler()
 
     def validate(self):
         """Check below requirements are met in setup.
 
-        1. required python 3rd party packages are installed
-        2. required rpm dependencies are installed
-        3. product specified in global config is supported by SSPL
+        1. Given product is supported by SSPL
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_post_install.py:59]: Check if given product ....
same for setup

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py [hunk 16]
```diff
@@ -20,72 +20,92 @@
...
-        # Make copy of global config
-        Conf.load(GLOBAL_CONFIG_INDEX, self.global_config_copy_url)
-        Conf.copy(self.prvsnr_global_config, GLOBAL_CONFIG_INDEX)
-        Conf.save(GLOBAL_CONFIG_INDEX)
+    name = "sspl_post_install"
+
+    def __init__(self):
+        """Initialize varibales for post install."""
+        self.user = "sspl-ll"
+        self.sspl_log_path = "/var/log/%s/sspl/" % consts.PRODUCT_FAMILY
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_post_install.py:51]: sspl_log_path 
sspl_bundle_path 

both these can be built in sspl_constants and used directly here ?

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+            self.validate_nw_cable_connection(i_list)
+        self.validate_nw_interfaces()
+
+    def process(self):
+        """Configure SSPL at prepare stage."""
+        pass
+
+    def validate_nw_cable_connection(self, interfaces):
+        """Check network interface links are up.
+        carrier:0 - link down
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:115]: 0 - cable disconnected
1- cable connected

link up/down can be because of reasons beyond cable connection

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py [hunk 18]
```diff
@@ -21,130 +21,56 @@
...
-            for item in files:
-                os.chown(os.path.join(root, item), uid, grpid)
+        """Initialize variables for sspl init."""
+        self.user = "sspl-ll"
+        Conf.load(consts.SSPL_CONFIG_INDEX, consts.sspl_config_path)
+
+    def validate(self):
+        """Check for below requirement.
+
+        1. Validate sspl-ll user exists
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:46]: validate if sspl-ll user exists

File: low-level/framework/base/sspl_constants.py [hunk 19]
```diff
@@ -66,9 +65,10 @@
 SSPL_LOGROTATE_CONF = "%s/sspl_logs" % LOGROTATE_DIR
 HPI_PATH = '/tmp/dcs/hpi'
 MDADM_PATH = '/etc/mdadm.conf'
-GLOBAL_CONFIG_INDEX = "global"
-SSPL_CONFIG_INDEX = "sspl"
-SSPL_TEST_CONFIG_INDEX = "sspl_test"
+PRVSNR_CONFIG_INDEX = "prvsnr_global_config"
```
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:68]: Call it prvsnr_input_config

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.yaml]: Now this added in sspl_constants.sh.

[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.yaml]: Done.

File: low-level/files/opt/seagate/sspl/conf/setup.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/setup.yaml:16]: Done.

File: low-level/actuators/impl/generic/node_hw.py [hunk 20]
```diff
@@ -26,8 +26,10 @@
 from actuators.impl.actuator import Actuator
 from framework.base.debug import Debug
 from framework.utils.service_logging import logger
-from framework.base.sspl_constants import AlertTypes, SensorTypes, SeverityTypes, COMMON_CONFIGS
-from framework.utils.conf_utils import Conf, CLUSTER, GLOBAL_CONF, SITE_ID, RACK_ID, NODE_ID, SRVNODE
+from framework.base.sspl_constants import (AlertTypes, SensorTypes,
+                                           SeverityTypes, COMMON_CONFIGS)
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/actuators/impl/generic/node_hw.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/generate_resource_health_view/resource_health_view (hunk 8 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/generate_resource_health_view/resource_health_view:41]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_config.py (hunk 9 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_config.py]: Removed assignment. But validation purpose we still keep this line.
Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py (hunk 10 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_post_install.py]: Removed.
Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py (hunk 11 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_post_install.py]: Removed.
Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py (hunk 12 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_post_install.py]: Removed.
Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 13 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py]: Removed.
Done.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml (hunk 14 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml]: Removed. Now we refer the hardcode value from sspl_constants.py
Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py (hunk 15 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_post_install.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_post_install.py (hunk 16 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_post_install.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 17 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py (hunk 18 above)
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py]: Done.

File: low-level/framework/base/sspl_constants.py (hunk 19 above)
Comment (mariyappanp) [reply low-level/framework/base/sspl_constants.py]: Done.

**Review by thavanathan — APPROVED** [review]

File: low-level/sspl-ll.spec (hunk 5 above)
Comment (cdeshmukh) [reply low-level/sspl-ll.spec:137]: rt, we can ignore sspl_init script as its unused from long now. Infact plan to remove it in coming days.

**Review by cdeshmukh — APPROVED** [review]
